extensions/s3/src/main/java/io/deephaven/extensions/s3/S3SeekableByteChannel.java

```diff
@@ -0,0 +1,321 @@
+            size = -1;
+        }
+
+        private int getIndex(final int fragmentIndex) {
+            return fragmentIndex % bufferCache.size();
+        }
+
+        void setFragmentContext(final int fragmentIndex, final CompletableFuture<ByteBuffer> future) {
+            final int cacheIdx = getIndex(fragmentIndex);
+            final FragmentContext cachedEntry = bufferCache.get(cacheIdx);
+            if (cachedEntry == null) {
+                bufferCache.set(cacheIdx, new FragmentContext(fragmentIndex, future));
+            } else {
+                // We should not cache an already cached fragment
+                Assert.neq(cachedEntry.fragmentIndex, "cachedEntry.fragmentIndex", fragmentIndex, "fragmentIdx");
+
+                // Cancel any outstanding requests for this cached fragment
+                cachedEntry.future.cancel(true);
+
+                // Reuse the existing entry
+                cachedEntry.fragmentIndex = fragmentIndex;
+                cachedEntry.future = future;
+            }
+        }
+
+        /**
+         * Will return the {@link CompletableFuture} corresponding to provided fragment index if present in the cache,
+         * else will return {@code null}
+         */
+        @Nullable
+        CompletableFuture<ByteBuffer> getCachedFuture(final int fragmentIndex) {
+            final FragmentContext cachedFragment = bufferCache.get(getIndex(fragmentIndex));
+            if (cachedFragment != null && cachedFragment.fragmentIndex == fragmentIndex) {
+                return cachedFragment.future;
+            }
+            return null;
+        }
+
+        long getSize() {
+            return size;
+        }
+
+        void setSize(final long size) {
+            this.size = size;
+        }
+    }
+
+    private final S3AsyncClient s3AsyncClient;
+    private final String bucket;
+    private final String key;
+    private final int fragmentSize;
+    private final int readAheadCount;
+    private final Duration readTimeout;
+
+    /**
+     * The size of the object in bytes, fetched at the time of first read
+     */
+    private long size;
+    private int numFragmentsInObject;
+
+    private S3ChannelContext context;
+
+    private long position;
+
+    S3SeekableByteChannel(@NotNull final SeekableChannelsProvider.ChannelContext context, @NotNull final URI uri,
+            @NotNull final S3AsyncClient s3AsyncClient, final int fragmentSize, final int readAheadCount,
+            final Duration readTimeout) {
+        final S3Uri s3Uri = s3AsyncClient.utilities().parseUri(uri);
+        this.bucket = s3Uri.bucket().orElse(null);
+        this.key = s3Uri.key().orElse(null);
+        this.s3AsyncClient = s3AsyncClient;
+        this.fragmentSize = fragmentSize;
+        this.readAheadCount = readAheadCount;
+        this.readTimeout = readTimeout;
+        Assert.instanceOf(context, "context", S3ChannelContext.class);
+        this.context = (S3ChannelContext) context;
+        this.size = -1;
```

Comment:
I would use an explicit `UNINITIALIZED_SIZE` constant. Same for the `S3ChannelContext`.